Radius API Guidelines document (#66)
radius-project/design-notes: guide/api-design-guidelines.md

The overall design of the Radius API is based on the Azure Resource Manager API (ARM). Radius generalizes the design of ARM, removes proprietary Azure concepts, and extends the ARM contract to support the conventions of other resource managers like Kubernetes or AWS Cloud-Control. 

Radius utilizes TypeSpec for describing API model definitions and operations, providing a consistent approach to defining the structure of API payloads and interactions. The TypeSpec specification serves as the source to generate corresponding OpenAPI 2.0 (swagger) API documentation.

Reference: 
- [Radius API](https://docs.radapp.io/concepts/technical/api/)
- [Radius Architecture](https://docs.radapp.io/concepts/technical/architecture/)

<a href="#secrets" name="secrets"></a>
### SECRETS

Radius enables users to securely store sensitive data, such as passwords, OAuth tokens, and SSH keys, in [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/).
The secrets are stored in secret store resource [Applications.Core/secretStores](https://docs.radapp.io/reference/resource-schema/core-schema/secretstore/)

> `SecretReference` and `SecretConfig` are entities prescribed to ensure sensitive information is handled securely and consistently across different components and resources in Radius. The following are their definitions along with usage examples:

Reviewer: ```suggestion
Our TypeSpec definitions provide the `SecretReference` and `SecretConfig` model types to standardize usage. These are prescribed to ensure sensitive information is handled securely and consistently across different components and resources in Radius. The following are their definitions along with usage examples:
```

### SecretReference Model

<a href="#secret-model" name="secret-model">:white_check_mark:</a> **DO** follow this structure when adding support for secrets to resources or components in Radius.

```

@doc("This specifies a reference to a secret. Secrets are encrypted, often have fine-grained access control, auditing and are recommended to be used to hold sensitive data.")
model SecretReference {
  @doc("The ID of an Applications.Core/SecretStore resource containing sensitive data.")
  source: string;

  @doc("The key for the secret in the secret store.")
  key: string;
}

```        

#### Usage of SecretReference in EnvironmentVariables:

> <a href="#secret-envvar" name="secret-envvar">:white_check_mark:</a> **DO** The above model for Secrets should be utilized for handling sensitive information through environment variables. 

Reviewer: Could we generalize this further? Something like:

> :white_check_mark: **DO** use the `SecretReference` type to reference a single (scalar) secret value in a resource property. Resource properties should reference a `Applications.Core/secretStores` instead of directly containing secret data.
>
> This pattern simplifies the overall design of Radius by reducing the number of places where we store secret data.

```

@doc("environment")
env?: Record<EnvironmentVariable>;

@doc("Environment variables type")
model EnvironmentVariable {

  @doc("The value of the environment variable")
  value?: string;

  @doc("The reference to the variable")
  valueFrom?: EnvironmentVariableReference;
}

@doc("The reference to the variable")
model EnvironmentVariableReference {
  @doc("The secret reference")
  secretRef: SecretReference;
}

```

#### Example in Bicep

```bicep

env: {
  DB_USER: { value: 'DB_USER' }
  DB_PASSWORD: {
    valueFrom: {
      secretRef: {
        source: secret.id
        key: 'DB_PASSWORD'
      }
    } 
  }
} 

```

### SecretConfig Model

<a href="#secretconfig-model" name="secretconfig-model">:white_check_mark:</a> **DO** follow this structure when a component or resource in Radius requires authentication to external systems, such as private container registries, TLS certificates.

```

@doc("Secret Configuration used to authenticate to external systems.")
model SecretConfig {
  @doc("The ID of an Applications.Core/secretStores resource containing credential information.")
  secret?: string;
}

```        

#### Using SecretConfig for Git repository access

> <a href="#secretconfig-ext" name="secretconfig-ext">:white_check_mark:</a> **DO** In the following example SecretConfig is used to manage authentication for accessing private Terraform modules from Git repository sources.

Reviewer: Could we generalize this further? Something like:

> :white_check_mark: **DO** use the `SecretConfig` type to reference a structured set of secret values in a resource property. Resource properties should reference a `Applications.Core/secretStores` instead of directly containing secret data.
>
> This pattern simplifies the overall design of Radius by reducing the number of places where we store secret data.

# Radius API Guidelines

<!-- markdownlint-disable MD033 MD049 MD055 -->
<!--
Note to contributors: All guidelines have an anchor tag to allow cross-referencing from associated tooling.
The anchor tags within a section using a common prefix to ensure uniqueness with anchor tags in other sections.
Please ensure that you add an anchor tag to any new guidelines that you add and maintain the naming convention.
-->

## History

<details>
  <summary>Expand change history</summary>

| Date        | Notes                                                          |
| ----------- | -------------------------------------------------------------- |
| 2024-Sep-17 | Added guidance on Secrets                                      |

</details>

## Introduction

These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to ensure that customers have a great experience. These guidelines help make Radius APIs:
- Developer friendly via consistent patterns

> - Efficient & cost-effective

Reviewer 2: Cost-effective as in **using less resources by following these patterns**?

Author: removed this sentence

Technology and software is constantly changing and evolving, and as such, this is intended to be a living document. [Open an issue](https://github.com/radius-project/design-notes/issues) to suggest a change or propose a new idea. 

> ### Prescriptive Guidance

Reviewer 3: I'm wondering if simplifying it down to three tiers of guidance might be more effective? i.e. DO, MAYBE, DO NOT

Author: discussed: keeping existing layers. remove language on exceptions from DO and DO NOT

This document offers prescriptive guidance labeled as follows:

:white_check_mark: **DO** adopt this pattern. If you feel you need an exception, present your reason at a design review discussion **prior** to implementation.

:ballot_box_with_check: **YOU SHOULD** adopt this pattern. If not following this advice, you MUST disclose your reason during a design review discussion.

:heavy_check_mark: **YOU MAY** consider this pattern if appropriate to your situation.

:warning: **YOU SHOULD NOT** adopt this pattern. If not following this advice, you MUST disclose your reason during a design review discussion.

:no_entry: **DO NOT** adopt this pattern. If you feel you need an exception, present your reason at a design review discussion **prior** to implementation.

## API Foundation:

Radius provides an HTTP-based API that deploys and manages cloud-native applications as well as on-premise or cloud resources.

> The overall design of the Radius API is based on the Azure Resource Manager API (ARM). Radius generalizes the design of ARM, removes proprietary Azure concepts, and extends the ARM contract to support the conventions of other resource managers like Kubernetes or AWS Cloud-Control. 

Reviewer 4: I'm not sure if `resource manager` has a standard meaning across users of different platforms?

```suggestion
The overall design of the Radius API is based on the Azure Resource Manager API (ARM). Radius generalizes the design of ARM, removes proprietary Azure concepts, and extends the ARM contract to support the conventions of other platforms like Kubernetes or AWS Cloud-Control. 
```

Author: discussed: we're keeping existing language

Radius utilizes TypeSpec for describing API model definitions and operations, providing a consistent approach to defining the structure of API payloads and interactions. The TypeSpec specification serves as the source to generate corresponding OpenAPI 2.0 (swagger) API documentation.

Reference: 
- [Radius API](https://docs.radapp.io/concepts/technical/api/)
- [Radius Architecture](https://docs.radapp.io/concepts/technical/architecture/)

## Guidelines

> ### Resource Property Design

Reviewer 5: These are all secret-related. Is the expectation that this file will be updated as more API patterns are used?

<a href="#secrets" name="secrets"></a>
### Secrets

Radius enables users to securely store sensitive data, such as passwords, OAuth tokens, and SSH keys, in [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/).

> The secrets are stored in secret store resource [Applications.Core/secretStores](https://docs.radapp.io/reference/resource-schema/core-schema/secretstore/)

Reviewer 4: `securely store sensitive data .... in [Kubernetes secrets]`

`The secrets are stored in secret store resource [Applications.Core/secretStores]`

Someone new to Radius might not understand the difference here. Applications.Core/secretStores is an abstraction whih we should make explicit.

Author: updated per discussion.

> Our TypeSpec definitions provide the `SecretReference` and `SecretConfig` model types to standardize usage. These are prescribed to ensure sensitive information is handled securely and consistently across different components and resources in Radius. The following are their definitions along with usage examples:

Reviewer 2: Should we add a link to the most recent **SecretReference** and **SecretConfig** in the code? Asking in case if there is an update and if we forget to update the models in this document below.

### SecretReference Model

<a href="#secret-model" name="secret-model">:white_check_mark:</a> **DO** follow this structure when adding support for secrets to resources or components in Radius.

> ```

Reviewer 6: can we add typespec syntax highlighting to these code blocks?

@doc("This specifies a reference to a secret. Secrets are encrypted, often have fine-grained access control, auditing and are recommended to be used to hold sensitive data.")
model SecretReference {
  @doc("The ID of an Applications.Core/SecretStore resource containing sensitive data.")
  source: string;

  @doc("The key for the secret in the secret store.")
  key: string;
}

```        

#### Usage of SecretReference

<a href="#secret-envvar" name="secret-envvar">:white_check_mark:</a> **DO** use the `SecretReference` type to reference a single (scalar) secret value in a resource property. Resource properties should reference a `Applications.Core/secretStores` instead of directly containing secret data.

This pattern simplifies the overall design of Radius by reducing the number of places where we store secret data. 

> This solution is in alignment with Kubernetes design patterns. The structure also allows for environment variables to refer to other resources such as ConfigMaps, Pod Fields etc. in the future. 

Reviewer 4: > This solution is in alignment with Kubernetes design patterns.

What is the benefit of calling this out?

Author: removed sentence

> Examples include cases where environment variables containing sensitive information are injected into a container or used in a Terraform execution process.

Reviewer 2: We can create an example in one of our repositories for common use cases, such as setting up a container registry secret that provides the cluster with the necessary credentials to connect to the registry and pull images.

#### Example TypeSpec Definition

```

@doc("environment")
env?: Record<EnvironmentVariable>;

@doc("Environment variables type")
model EnvironmentVariable {

  @doc("The value of the environment variable")
  value?: string;

  @doc("The reference to the variable")
  valueFrom?: EnvironmentVariableReference;
}

@doc("The reference to the variable")
model EnvironmentVariableReference {
  @doc("The secret reference")
  secretRef: SecretReference;
}

```

#### Example Bicep Definition

```bicep

env: {
  DB_USER: { value: 'DB_USER' }
  DB_PASSWORD: {
    valueFrom: {
      secretRef: {
        source: secret.id
        key: 'DB_PASSWORD'
      }
    } 
  }
} 

```

### SecretConfig Model

> <a href="#secretconfig-model" name="secretconfig-model">:white_check_mark:</a> **DO** follow this structure when a component or resource in Radius requires authentication to external systems, such as private container registries, TLS certificates.

Reviewer 7: i think we can rephrase this into something like : "follow this structure when required keys are known"? e.g  secret store used for private terraform recipe config, we know the expected secret keys are `username` and `pat`

# Radius API Guidelines

<!-- markdownlint-disable MD033 MD049 MD055 -->
<!--
Note to contributors: All guidelines have an anchor tag to allow cross-referencing from associated tooling.
The anchor tags within a section using a common prefix to ensure uniqueness with anchor tags in other sections.
Please ensure that you add an anchor tag to any new guidelines that you add and maintain the naming convention.
-->

## History

<details>
  <summary>Expand change history</summary>

| Date        | Notes                                                          |
| ----------- | -------------------------------------------------------------- |
| 2024-Sep-17 | Added guidance on Secrets                                      |

</details>

## Introduction

> These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to ensure that customers have a great experience. These guidelines help make Radius APIs:

Reviewer 3: ```suggestion
These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs:
```

- Developer friendly via consistent patterns

Technology and software is constantly changing and evolving, and as such, this is intended to be a living document. [Open an issue](https://github.com/radius-project/design-notes/issues) to suggest a change or propose a new idea. 

### Prescriptive Guidance

> This document offers prescriptive guidance labeled as follows:

Reviewer 8: what is the difference between a `do adopt` and `may adopt`?  both left to the user to decide?

Author: Per discussion updated to 'YOU MUST' and 'YOU MUST NOT' ..

:warning: **YOU SHOULD NOT** adopt this pattern. If not following this advice, you MUST disclose your reason during a design review discussion.

:no_entry: **DO NOT** adopt this pattern.

## API Foundation:

Radius provides an HTTP-based API that deploys and manages cloud-native applications as well as on-premise or cloud resources.
The overall design of the Radius API is based on the Azure Resource Manager API (ARM). Radius generalizes the design of ARM, removes proprietary Azure concepts, and extends the ARM contract to support the conventions of other resource managers like Kubernetes or AWS Cloud-Control. 

Radius utilizes TypeSpec for describing API model definitions and operations, providing a consistent approach to defining the structure of API payloads and interactions. The TypeSpec specification serves as the source to generate corresponding OpenAPI 2.0 (swagger) API documentation.

Reference: 
- [Radius API](https://docs.radapp.io/concepts/technical/api/)
- [Radius Architecture](https://docs.radapp.io/concepts/technical/architecture/)

## Guidelines

### Resource Property Design

<a href="#secrets" name="secrets"></a>

> ### Secrets

Reviewer 2: Is this supposed to be under **Resource Property Design**? If so, you need 4 #'s here.

If not, should we add information to the **Resource Property Design** section?

<a href="#secret-store" name="secret-store">:white_check_mark:</a> **DO** use [Applications.Core/secretStores](https://docs.radapp.io/reference/resource-schema/core-schema/secretstore/) resource to store sensitive data, such as passwords, OAuth tokens, and SSH keys. Through the use of this resource, Radius ensures that confidential data is handled in a safe and consistent manner.

> 

Reviewer: Would it be less confusing to call out that there are two different patterns in this section? I'm just thinking about what would be easier for a reader to grok.

Something like:

> The applicable guidance for API design of secrets is slightly different depending on which of these two scenarios is the best fit:
>
> - Storing and retrieving a single (scalar) secret value. eg: An environment variable. 
> - Storing and retrieving a structured (multiple-value) secret. eg: OAuth configuration.

> The TypeSpec definitions include [`SecretReference`](https://github.com/radius-project/radius/blob/ba9195c4bc6ab9be2a83bb1ed6b0f8f357d79812/typespec/Applications.Core/environments.tsp#L193) and [`SecretConfig`](https://github.com/radius-project/radius/blob/ba9195c4bc6ab9be2a83bb1ed6b0f8f357d79812/typespec/Applications.Core/environments.tsp#L126) model types to standardize usage. These are prescribed to ensure sensitive information is handled securely and uniformly across different components and resources in Radius. The following are their definitions along with usage examples:

Reviewer 2: We give the link to a specific commit. Is this intended? What if this changes in a few months and the contributors who read this doc see the old version of the **SecretReference**?

Author: updated document. these links were removed. let's review again

> ### SecretReference Model

Reviewer: Should this be nested under `### Secrets`? It feels like this whole section deals with secrets, but the subheadings like `### SecretReference Model` are at the same level of nesting.

Author: Secrets is updated to 4 # as a heading and everything under it is now 5 #

```

#### Example Bicep Definition

```bicep

env: {
  DB_USER: { value: 'DB_USER' }
  DB_PASSWORD: {
    valueFrom: {
      secretRef: {
        source: secret.id
        key: 'DB_PASSWORD'
      }
    } 
  }
} 

```

### :no_entry: DO NOT: Store sensitive data as clear text
Sensitive data MUST NEVER be stored directly in resource definitions or configuration files as clear text. Instead, use references to secrets stored in secret stores.

#### Example of What NOT to Do

```bicep

env: {
  DB_USER: { value: 'DB_USER' }
  DB_PASSWORD:  { value: 'myPlainTextPassword' }  // ❌ This is NOT allowed
} 

```

### SecretConfig Model

<a href="#secretconfig-model" name="secretconfig-model">:white_check_mark:</a> **DO** implement this model when the structure of the secret store resource is known and a component or resource in Radius requires authentication to external systems, such as private container registries, TLS certificates.

```tsp

@doc("Secret Configuration used to authenticate to external systems.")
model SecretConfig {
  @doc("The ID of an Applications.Core/secretStores resource containing credential information.")
  secret?: string;
}

```        

#### Usage of SecretConfig

> <a href="#secretconfig-ext" name="secretconfig-ext">:white_check_mark:</a> **DO** use the SecretConfig type to reference a structured set of secret values in a resource property. Resource properties should reference a Applications.Core/secretStores instead of directly containing secret data.

Reviewer 3: ```suggestion
<a href="#secretconfig-ext" name="secretconfig-ext">:white_check_mark:</a> **DO** use the `SecretConfig` type to reference a structured set of secret values in a resource property. Resource properties should reference a `Applications.Core/secretStores` instead of directly containing secret data.
```

# Radius API Guidelines

<!-- markdownlint-disable MD033 MD049 MD055 -->
<!--
Note to contributors: All guidelines have an anchor tag to allow cross-referencing from associated tooling.
The anchor tags within a section using a common prefix to ensure uniqueness with anchor tags in other sections.
Please ensure that you add an anchor tag to any new guidelines that you add and maintain the naming convention.
-->

## History

<details>
  <summary>Expand change history</summary>

| Date        | Notes                                                          |
| ----------- | -------------------------------------------------------------- |
| 2024-Sep-17 | Added guidance on Secrets                                      |

</details>

## Introduction

> These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs developer friendly via consistent patterns.  
> Technology and software is constantly changing and evolving, and as such, this is intended to be a living document, with a current focus on secrets handling. [Open an issue](https://github.com/radius-project/design-notes/issues) to suggest a change or propose a new idea.

Reviewer 4: ```suggestion
These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs developer friendly via consistent patterns.

Technology and software is constantly changing and evolving, and as such, this is intended to be a living document, with a current focus on secrets handling. [Open an issue](https://github.com/radius-project/design-notes/issues) to suggest a change or propose a new idea.
```

# Radius API Guidelines

<!-- markdownlint-disable MD033 MD049 MD055 -->
<!--
Note to contributors: All guidelines have an anchor tag to allow cross-referencing from associated tooling.
The anchor tags within a section using a common prefix to ensure uniqueness with anchor tags in other sections.
Please ensure that you add an anchor tag to any new guidelines that you add and maintain the naming convention.
-->

## History

<details>
  <summary>Expand change history</summary>

| Date        | Notes                                                          |
| ----------- | -------------------------------------------------------------- |
| 2024-Sep-17 | Added guidance on Secrets                                      |

</details>

## Introduction

> These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs:

Reviewer 4: @lakshmimsft, To set the right expectations for readers of this document, I'd recommend being explicit here that currently the guidance is focused on secrets handling. cc @rynowak for inputs as well.

Author: updated

> These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs:
> - Developer friendly via consistent patterns

Reviewer 4: nit: do we need to separate it in bullet points given that there is only one?

```suggestion
These are prescriptive guidelines that Radius contributors MUST follow while designing APIs to maintain a great user experience. These guidelines help make Radius APIs developer friendly via consistent patterns
```

## Guidelines

### Resource Property Design

<a href="#secrets" name="secrets"></a>
#### Secrets

<a href="#secret-store" name="secret-store">:white_check_mark:</a> **YOU MUST** use [Applications.Core/secretStores](https://docs.radapp.io/reference/resource-schema/core-schema/secretstore/) resource to store sensitive data, such as passwords, OAuth tokens, and SSH keys. Through the use of this resource, Radius ensures that confidential data is handled in a safe and consistent manner.

The guidance for API design of secrets depends on which of these two scenarios best fits your use case:
- For storing and retrieving a single (scalar) secret value, e.g., an environment variable, refer to [`SecretReference`](#secretreference-model) model type described below.
- When storing and retrieving a structured (multiple-value) secret, e.g., OAuth configuration, refer to [`SecretConfig`](#secretconfig-model) model type described below.

 These prescribed types ensure secure and uniform handling of sensitive information across different components and resources in Radius. Their definitions and usage examples are as follows:
 
##### SecretReference Model

> <a href="#c name="secretreference-model">:white_check_mark:</a> **YOU MUST** follow this structure when adding support for secrets to resources or components in Radius.

Reviewer 4: I think there is a formatting issue here based on preview.
